// ngx-webstorage-service/esm5/src/storage-transcoders.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Transcoder that encodes values as JSON strings.
 */
var /**
 * Transcoder that encodes values as JSON strings.
 */
JsonStorageTranscoder = /** @class */ (function () {
    function JsonStorageTranscoder() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    JsonStorageTranscoder.prototype.encode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return JSON.stringify(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    JsonStorageTranscoder.prototype.decode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            return JSON.parse(value);
        }
        catch (error) {
            return undefined;
        }
    };
    return JsonStorageTranscoder;
}());
/**
 * Transcoder that encodes values as JSON strings.
 */
export { JsonStorageTranscoder };
/**
 * Transcoder that encodes/decodes strings **as is**, i.e. values are not modified in any way.
 */
var /**
 * Transcoder that encodes/decodes strings **as is**, i.e. values are not modified in any way.
 */
StringStorageTranscoder = /** @class */ (function () {
    function StringStorageTranscoder() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    StringStorageTranscoder.prototype.encode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    StringStorageTranscoder.prototype.decode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value;
    };
    return StringStorageTranscoder;
}());
/**
 * Transcoder that encodes/decodes strings **as is**, i.e. values are not modified in any way.
 */
export { StringStorageTranscoder };
/**
 * Transcoder that encodes/decodes `boolean` values.
 */
var /**
 * Transcoder that encodes/decodes `boolean` values.
 */
BooleanStorageTranscoder = /** @class */ (function () {
    function BooleanStorageTranscoder() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    BooleanStorageTranscoder.prototype.encode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value.toString();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    BooleanStorageTranscoder.prototype.decode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value === 'true') {
            return true;
        }
        if (value === 'false') {
            return false;
        }
        return undefined;
    };
    return BooleanStorageTranscoder;
}());
/**
 * Transcoder that encodes/decodes `boolean` values.
 */
export { BooleanStorageTranscoder };
/**
 * Transcoder that encodes/decodes `number` values.
 */
var /**
 * Transcoder that encodes/decodes `number` values.
 */
NumberStorageTranscoder = /** @class */ (function () {
    function NumberStorageTranscoder() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    NumberStorageTranscoder.prototype.encode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value.toString();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NumberStorageTranscoder.prototype.decode = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var parsedNumber = Number(value);
        return Number.isFinite(parsedNumber) ? parsedNumber : undefined;
    };
    return NumberStorageTranscoder;
}());
/**
 * Transcoder that encodes/decodes `number` values.
 */
export { NumberStorageTranscoder };
/**
 * A set of storage transcoders.
 * @type {?}
 */
export var StorageTranscoders = {
    /**
     * Transcoder that encodes values as JSON strings.
     */
    JSON: (/** @type {?} */ (new JsonStorageTranscoder())),
    /**
     * Transcoder that encodes/decodes strings **as is**, i.e. values are not modified in any way.
     */
    STRING: (/** @type {?} */ (new StringStorageTranscoder())),
    /**
     * Transcoder that encodes/decodes `boolean` values.
     */
    BOOLEAN: (/** @type {?} */ (new BooleanStorageTranscoder())),
    /**
     * Transcoder that encodes/decodes `number` values.
     */
    NUMBER: (/** @type {?} */ (new NumberStorageTranscoder()))
};
